From d8edc3e901ceee176c86613d1f423debfac0d0f3 Mon Sep 17 00:00:00 2001
From: Jeff <jeffswenson@betterthannull.com>
Date: Sun, 12 Jul 2015 14:47:29 -0500
Subject: [PATCH] Make changes needed for patch file.

---
 src/SFML/Main/MainAndroid.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/SFML/Main/MainAndroid.cpp b/src/SFML/Main/MainAndroid.cpp
index 52e3816..e4af0a1 100644
--- a/src/SFML/Main/MainAndroid.cpp
+++ b/src/SFML/Main/MainAndroid.cpp
@@ -117,7 +117,7 @@ void* main(ActivityStates* states)
     initializeMain(states);
 
     sleep(seconds(0.5));
-    ::main(0, NULL);
+    ::main(0, (char**)states->activity->assetManager);
 
     // Terminate properly the main thread and wait until it's done
     terminateMain(states);
-- 
2.4.5

