diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6750436..c98da3c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -71,8 +71,16 @@ else ()
   target_link_libraries(SDL2_ttf SDL2::SDL2-static)
   set_target_properties(SDL2_ttf PROPERTIES POSITION_INDEPENDENT_CODE ON)
 endif ()
-target_link_libraries(SDL2_ttf Freetype::Freetype)
+
+if(CMAKE_BUILD_TYPE STREQUAL "DEBUG")
+    set(LIB_SUFFIX "d")
+else()
+    set(LIB_SUFFIX "")
+endif()
+target_link_directories(SDL2_ttf PRIVATE ${RAFT_LIB_DIR})
+target_link_libraries(SDL2_ttf freetype${LIB_SUFFIX})
 target_include_directories(SDL2_ttf PUBLIC $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}/SDL2>)
+target_include_directories(SDL2_ttf PRIVATE "${RAFT_INCLUDE_DIR}/freetype2")
 
 install(
   TARGETS SDL2_ttf
@@ -83,7 +91,7 @@ install(
 )
 install(
   FILES SDL_ttf.h
-  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/SDL2
+  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/SDL2_ttf
 )
 
 ##### export files #####
