diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6750436..e8d49c5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -83,7 +83,7 @@ install(
 )
 install(
   FILES SDL_ttf.h
-  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/SDL2
+  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/SDL2_ttf
 )
 
 ##### export files #####
@@ -114,6 +114,12 @@ install(
   COMPONENT Devel
 )
 
+if (BUILD_SHARED_LIBS)
+	if(WIN32)
+	  target_compile_definitions(SDL2_ttf PRIVATE -DDLL_EXPORT)
+	endif()
+endif()
+
 ##### pkg-config #####
 if (PKG_CONFIG_FOUND)
   set(SDL_VERSION 2.0.0)
