diff --git a/src/SFML/Window/iOS/SFViewController.hpp b/src/SFML/Window/iOS/SFViewController.hpp
index 65b6eda0..fcff6a60 100644
--- a/src/SFML/Window/iOS/SFViewController.hpp
+++ b/src/SFML/Window/iOS/SFViewController.hpp
@@ -63,6 +63,8 @@
 ////////////////////////////////////////////////////////////
 - (NSUInteger)supportedInterfaceOrientations;
 
+- (void)viewDidAppear:(bool)animated;
+
 ////////////////////////////////////////////////////////////
 // Member data
 ////////////////////////////////////////////////////////////
diff --git a/src/SFML/Window/iOS/SFViewController.mm b/src/SFML/Window/iOS/SFViewController.mm
index 582b3ff1..f66a70bd 100644
--- a/src/SFML/Window/iOS/SFViewController.mm
+++ b/src/SFML/Window/iOS/SFViewController.mm
@@ -53,5 +53,14 @@ - (NSUInteger)supportedInterfaceOrientations
     return UIInterfaceOrientationMaskAll;
 }
 
+- (void)viewDidAppear:(bool)animated
+{
+    auto window = self.view.window;
+    auto gr0 = window.gestureRecognizers[0];
+    auto gr1 = window.gestureRecognizers[1];
+    gr0.delaysTouchesBegan = false;
+    gr1.delaysTouchesBegan = false;
+}
+
 
 @end
