diff --git a/src/SFML/Window/iOS/SFViewController.hpp b/src/SFML/Window/iOS/SFViewController.hpp
index 65b6eda0..61f0e025 100644
--- a/src/SFML/Window/iOS/SFViewController.hpp
+++ b/src/SFML/Window/iOS/SFViewController.hpp
@@ -29,13 +29,14 @@
 // Headers
 ////////////////////////////////////////////////////////////
 #include <UIKit/UIKit.h>
 
 
 ////////////////////////////////////////////////////////////
 /// \brief The view controller handles the view's orientation
 ///
 ////////////////////////////////////////////////////////////
 
 ////////////////////////////////////////////////////////////
 /// \brief Tells if the controller supports auto-rotation (iOS < 6)
@@ -63,11 +64,15 @@
 ////////////////////////////////////////////////////////////
 - (NSUInteger)supportedInterfaceOrientations;
 
+- (void)viewDidAppear:(bool)animated;
+
 ////////////////////////////////////////////////////////////
 // Member data
 ////////////////////////////////////////////////////////////
 @property (nonatomic) bool orientationCanChange; ///< Tells whether the controller's view can rotate or not
 
+
 @end
 
 #endif // SFML_SFVIEWCONTROLLER_HPP
diff --git a/src/SFML/Window/iOS/SFViewController.mm b/src/SFML/Window/iOS/SFViewController.mm
index 582b3ff1..246a8188 100644
--- a/src/SFML/Window/iOS/SFViewController.mm
+++ b/src/SFML/Window/iOS/SFViewController.mm
@@ -50,8 +50,24 @@ - (BOOL)shouldAutorotate
 ////////////////////////////////////////////////////////////
 - (NSUInteger)supportedInterfaceOrientations
 {
-    return UIInterfaceOrientationMaskAll;
+    //return UIInterfaceOrientationMaskAll;
+    return UIInterfaceOrientationMaskLandscapeLeft | UIInterfaceOrientationMaskLandscapeRight;
 }
 
+- (void)viewDidAppear:(bool)animated
+{
+    auto window = self.view.window;
+    auto gr0 = window.gestureRecognizers[0];
+    auto gr1 = window.gestureRecognizers[1];
+    gr0.delaysTouchesBegan = false;
+    gr1.delaysTouchesBegan = false;
+}
 
 @end
